### PR TITLE
8334495: Use FFM instead of jdk.internal.misc.Unsafe in java.desktop font implementation

### PR DESCRIPTION
**Pull request by prrace**
Migrate font code from jdk.internal.misc.Unsafe to using FFM.
This reduces the coupling between the java.desktop module and the internals of the java.base module.

The code being changed here is not particularly performance sensitive, and it is not executed in the most common cases.
The main impact performance-wise is a total of around 37ms in initialisation costs on my x64 macbook.
A minimal program that just draws a string to an image - does not even put up a window - runs at around 690-700ms.
There's variability in that number and the overall time for a JDK without the change is around (660-670ms)
In the small test, this is the first and only use of FFM, so the one-off part cost should move elsewhere when FFM starts
to be used earlier in the JDK itself.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8334495](https://bugs.openjdk.org/browse/JDK-8334495): Use FFM instead of jdk.internal.misc.Unsafe in java.desktop font implementation (**Bug** - P4)

### Reviewers
 * [Damon Nguyen](https://openjdk.org/census#dnguyen) (@DamonGuy - Committer) 🔄 Re-review required (review applies to [aa566257](https://git.openjdk.org/jdk/pull/19777/files/aa56625788c85ee0878c0efbe49127baa3711681))
 * [Alisen Chung](https://openjdk.org/census#achung) (@alisenchung - Committer) 🔄 Re-review required (review applies to [99acf847](https://git.openjdk.org/jdk/pull/19777/files/99acf84717d4bc4459ba646a7c130ec8f56b1a34))
 * [Jayathirth D V](https://openjdk.org/census#jdv) (@jayathirthrao - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19777/head:pull/19777` \
`$ git checkout pull/19777`

Update a local copy of the PR: \
`$ git checkout pull/19777` \
`$ git pull https://git.openjdk.org/jdk.git pull/19777/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19777`

View PR using the GUI difftool: \
`$ git pr show -t 19777`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19777.diff">https://git.openjdk.org/jdk/pull/19777.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19777#issuecomment-2176927639)

### CODE REVIEW

**Comment by alisenchung** [comment]
How do I test this fix?

**Review by DamonGuy — APPROVED** [review]
Changes look good as far as I can tell. I tested it on Ubuntu 22.04 on SwingSet2. Everything working as expected. Can't give any exact details on performance, but it runs smoothly as far as I can tell. No major difference performance-wise, which I suppose is a good thing.

File: src/java.desktop/share/classes/sun/font/StrikeCache.java [hunk 1]
```diff
@@ -98,40 +105,161 @@ public final class StrikeCache {
...
+    private static final VarHandle heightHandle   = getVarHandle(GlyphImageLayout, "height");
+    private static final VarHandle rowBytesHandle = getVarHandle(GlyphImageLayout, "rowBytes");
+    private static final VarHandle managedHandle  = getVarHandle(GlyphImageLayout, "managed");
+    private static final VarHandle topLeftXHandle = getVarHandle(GlyphImageLayout, "topLeftX");
+    private static final VarHandle topLeftYHandle = getVarHandle(GlyphImageLayout, "topLeftY");
+    private static final VarHandle cellInfoHandle  = getVarHandle(GlyphImageLayout, "cellInfo");
+    private static final VarHandle imageHandle    = getVarHandle(GlyphImageLayout, "image");
+
+    @SuppressWarnings("restricted")
+    static final float getGlyphXAdvance(long ptr) {
```
Comment (mcimadamore) [inline src/java.desktop/share/classes/sun/font/StrikeCache.java:151]: now, I'm not an expert of this code, but I notice that you have accessors that seem to take a bare `long` instead of `MemorySegment`. Have you tried pushing segments deeper in the implementation? That way I think you could completely auto-generate this code using jextract. (of course what you have is not bad - I'm mostly trying to see if there's a way to get there w/o all these calls to `MemorySegment::reinterpret`).

Comment (mcimadamore) [reply src/java.desktop/share/classes/sun/font/StrikeCache.java:151]: To be clear - I'm assuming the `ptr` parameter comes from a single native call. That call returns a pointer to a struct, or maybe an array of structs. If so, we could reinterpret right after the native call, with the right size (the number of structs returned by the native call), and then all the accessors will be just "plain" memory segment accessors, no need to reinterpret (as the segment will already have the correct size).

Comment (prrace) [reply src/java.desktop/share/classes/sun/font/StrikeCache.java:151]: Each pointer refers to a single struct, not an array.
In this fix I am focused on not using Unsafe.
I don't want the change to go further than that.
The 'long's are retrieved/ passed around/ back down to native in lots of places so it would turn a  focused change into a much bigger change.
grep in java.desktop for "getGlyphImage" to see the tip of the iceberg.

File: src/java.desktop/share/classes/sun/font/StrikeCache.java [hunk 2]
```diff
@@ -98,40 +105,161 @@ public final class StrikeCache {
...
+        MemorySegment seg = MemorySegment.ofAddress(ptr);
+        seg = seg.reinterpret(GLYPHIMAGESIZE);
+        char hgt = (char)heightHandle.get(seg);
+        char rb = (char)rowBytesHandle.get(seg);
+        MemorySegment pixelData = (MemorySegment)imageHandle.get(seg);
+        int sz = rb * hgt;
+        pixelData = pixelData.reinterpret(sz);
+        byte[] bytes = new byte[sz];
+        MemorySegment.copy(pixelData, ValueLayout.JAVA_BYTE, 0, bytes, 0, sz);
+        return bytes;
```
Comment (JornVernee) [inline src/java.desktop/share/classes/sun/font/StrikeCache.java:257]: You could use `toArray` here:

```suggestion
        return pixelData.toArray(ValueLayout.JAVA_BYTE);
```

**Comment by mrserb** [comment]
DId ffm team check from where that slow down come from? is it expected to lose 5%?

>In the small test, this is the first and only use of FFM, so the one-off part cost should move elsewhere when FFM starts
to be used earlier in the JDK itself.

So the changed code path will work as before (w/o slowdown) if the https://github.com/openjdk/jdk/pull/15476 will be touched by the test? and slow down will occur if "sun.font.layout.ffm" will be set to false? Just would like to confirm that we will not get "multiplication of slowdowns" for each ffm usage.

File: src/java.desktop/share/classes/sun/font/StrikeCache.java [hunk 3]
```diff
@@ -98,40 +105,159 @@ public final class StrikeCache {
...
-     * arr[4] = offset of width
-     * arr[5] = offset of height
-     * arr[6] = offset of rowBytes
-     * arr[7] = offset of topLeftX
-     * arr[8] = offset of topLeftY
-     * arr[9] = offset of pixel data.
-     * arr[10] = address of a GlyphImageRef representing the invisible glyph
-     */
-    static native void getGlyphCacheDescription(long[] infoArray);
+    static final int nativeAddressSize = (int)ValueLayout.ADDRESS.byteSize();
```
Comment (minborg) [inline src/java.desktop/share/classes/sun/font/StrikeCache.java:112]: Suggestion: NATIVE_ADDRESS_SIZE

**Comment by prrace** [comment]
> DId ffm team check from where that slow down come from? is it expected to lose 5%?
> 
> > In the small test, this is the first and only use of FFM, so the one-off part cost should move elsewhere when FFM starts
> > to be used earlier in the JDK itself.
> 
> So the changed code path will work as before (w/o slowdown) if the #15476 will be touched by the test? and slow down will occur if "sun.font.layout.ffm" will be set to false? Just would like to confirm that we will not get "multiplication of slowdowns" for each ffm usage.

Well there's going to be some fraction of it in order to create the VarHandles but
the initialisation cost of the classes used by the FFM implementation is
one time even if both are used, so it just depends who gets there first as to where
it is seen, and if both are used, then only one will take that initialisation cost.
However the layout code creates downcall and upcall stubs too, so it has a bigger
one time cost than this code.
In other words if hypothetically :
- 30 ms one time init machinery cost for this cache code
- 80 ms one time init machinery cost for the layout FFM code
and this code is executed first, then layout would see a 50 ms cost,
or if the layout code is executed first then I'd expect to see zero machinery init cost here.

There's still *some* cost for each VarHandle etc. Executing code isn't free.

File: src/java.desktop/unix/classes/sun/font/XRGlyphCacheEntry.java [hunk 4]
```diff
@@ -69,28 +69,25 @@ public int getGlyphSet() {
...
+        // We need to access the GlyphID as a long because the
         // corresponding field in the underlying C data-structure is of type
-        // 'void*' (see field 'cellInfo' of struct 'GlyphInfo' in
-        // src/share/native/sun/font/fontscalerdefs.h).
-        // On 64-bit Big-endian architectures it would be wrong to write this
-        // field with Unsafe.putInt() because it is also accessed from native
-        // code as a 'long'.
+        // 'void*' (see field 'cellInfo' of struct 'GlyphInfo'
+        // in src/share/native/sun/font/fontscalerdefs.h).
+        // On 64-bit Big-endian architectures it would be wrong to access this
```
Comment (alisenchung) [inline src/java.desktop/unix/classes/sun/font/XRGlyphCacheEntry.java:86]: "wrong to write this field as an int"?

**Review by alisenchung — APPROVED** [review]
Just some suggestions on comments

File: src/java.desktop/unix/classes/sun/font/XRGlyphCacheEntry.java (hunk 4 above)
Comment (prrace) [reply src/java.desktop/unix/classes/sun/font/XRGlyphCacheEntry.java:86]: "access" because it is true for read and write.

File: src/java.desktop/share/classes/sun/font/GlyphList.java [hunk 5]
```diff
@@ -373,9 +364,8 @@ public byte[] getGrayBits() {
          * Consider a native method if there is a performance problem (which
```
[the remark this answers is not in the document]
Comment (prrace) [reply src/java.desktop/share/classes/sun/font/GlyphList.java:364]: Yes, in fact I thought I HAD removed it. Thanks for spotting.

File: src/java.desktop/share/classes/sun/font/StrikeCache.java [hunk 6]
```diff
@@ -98,40 +105,159 @@ public final class StrikeCache {
...
+
+    private static final VarHandle xAdvanceHandle = getVarHandle(GlyphImageLayout, "xAdvance");
+    private static final VarHandle yAdvanceHandle = getVarHandle(GlyphImageLayout, "yAdvance");
+    private static final VarHandle widthHandle    = getVarHandle(GlyphImageLayout, "width");
+    private static final VarHandle heightHandle   = getVarHandle(GlyphImageLayout, "height");
+    private static final VarHandle rowBytesHandle = getVarHandle(GlyphImageLayout, "rowBytes");
+    private static final VarHandle managedHandle  = getVarHandle(GlyphImageLayout, "managed");
+    private static final VarHandle topLeftXHandle = getVarHandle(GlyphImageLayout, "topLeftX");
+    private static final VarHandle topLeftYHandle = getVarHandle(GlyphImageLayout, "topLeftY");
+    private static final VarHandle cellInfoHandle  = getVarHandle(GlyphImageLayout, "cellInfo");
```
Comment (turbanoff) [inline src/java.desktop/share/classes/sun/font/StrikeCache.java:147]:
```suggestion
    private static final VarHandle cellInfoHandle = getVarHandle(GlyphImageLayout, "cellInfo");
```

**Comment by mrserb** [comment]
> or if the layout code is executed first then I'd expect to see zero machinery init cost here.

It will be good to check thi.s So we'll use it freely, assuming it won't introduce any additional overhead.

**Comment by prrace** [comment]
> > or if the layout code is executed first then I'd expect to see zero machinery init cost here.
> 
> It will be good to check thi.s So we'll use it freely, assuming it won't introduce any additional overhead.

StrikeCache will always be initialised before you can execute the FFM-based harfbuzz code.
So all I can measure is how much faster does the FFM HB get (first time) as a result of the pre-initialisation of FFM.

So I have 2 JDK builds, one with both changes, the other with just the pre-existing harfbuzz changes.
In both I instrumented around the SunLayoutEngine.java code that initialises HBShaper.java
to record the time to execute the first call to layout.

In the case where ONLY the FFM harfbuzz changes are in it is around 178ms (best result of 10)
In the case where BOTH FFM changes are in it is around 123 ms (best result of 10)
There were quite a few of the runs in that ballpark
So it is around 55ms faster to run the hb code after already initialising the FFM-using StrikeCache.

NB those times (124ms and 178 ms ) include the execution time for layout as well,
but the difference is what matters to answer this question and it is answered as expected.

**Review by jayathirthrao — APPROVED** [review]
Change looks good to me.
Its good that we have captured the time for computation when we have FFM usage at more than one place and how FFM initialization overhead is hit only once.

Also in https://github.com/openjdk/jdk/pull/15476 we had FFM initialization cost of 75ms and now we are seeing 55ms number. If same hardware & software conditions were present between https://github.com/openjdk/jdk/pull/15476 and this PR, may be we are moving towards lower FFM initialization cost.

**Comment by mrserb** [comment]
> NB those times (124ms and 178 ms ) include the execution time for layout as well,
> but the difference is what matters to answer this question and it is answered as expected.

This is a good news!
